src/controllers/krate/versions.rs

```diff
@@ -302,4 +354,228 @@ struct PaginatedVersionsAndPublishers {
 struct ResponseMeta {
     total: i64,
     next_page: Option<String>,
+    release_tracks: Option<ReleaseTracks>,
+}
+
+#[derive(Debug, Eq, PartialEq, Serialize)]
+struct ReleaseTracks(IndexMap<ReleaseSlug, HighestSemver>);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+enum ReleaseSlug {
+    MajorX(u64),
+    MajorMinor(u64, u64),
+    Major(u64),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
+struct HighestSemver {
+    highest: SemverTriple,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct SemverTriple(u64, u64, u64);
+
+impl ReleaseTracks {
+    // Return the release tracks based on a sorted semver versions iterator (in descending order).
+    // **Remember to** filter out yanked versions manually before calling this function.
+    pub fn from_sorted_semver_iter<'a, I>(versions: I) -> Self
+    where
+        I: Iterator<Item = &'a semver::Version>,
+    {
+        let mut map = IndexMap::new();
+        for num in versions.filter(|num| num.pre.is_empty()) {
+            let key = ReleaseSlug::from_semver(num);
+            let prev = map.last();
+            if prev.filter(|&(k, _)| *k == key).is_none() {
+                map.insert(key, HighestSemver::from_semver(num));
+            }
+        }
+
+        Self(map)
+    }
+}
+
+impl ReleaseSlug {
+    pub fn from_semver(version: &semver::Version) -> Self {
+        if version.major >= 100 {
+            Self::Major(version.major)
+        } else if version.major == 0 {
+            Self::MajorMinor(version.major, version.minor)
+        } else {
+            Self::MajorX(version.major)
+        }
+    }
+}
+
+impl HighestSemver {
+    pub fn from_semver(version: &semver::Version) -> Self {
+        Self {
+            highest: SemverTriple(version.major, version.minor, version.patch),
+        }
+    }
+}
+
+impl std::fmt::Display for ReleaseSlug {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::MajorX(major) => write!(f, "{major}.x"),
+            Self::MajorMinor(major, minor) => write!(f, "{major}.{minor}"),
+            Self::Major(major) => write!(f, "{major}"),
```

Comment:
I'm wondering if we should really encode the `.x` thing in the backend. It's more of a presentational thing. in the `MajorMinor` case `major` is always zero, so we might not need to store the zero in the enum.

Reply:
> I'm wondering if we should really encode the `.x` thing in the backend. It's more of a presentational thing.

I just implemented it to match the current frontend's implementation.
Do we want to change the representation in this PR or just change it in a follow-up PR if necessary?



> in the `MajorMinor` case `major` is always zero, so we might not need to store the zero in the enum.

Great call!

Reply:
I think I got your point now. Are you suggesting that instead of providing `1.x` we could simply provide `1` and let client to determine what to display? In that case, I think two variants `Major` and `Minor` would be sufficient.

Reply:
yeah, exactly